File: frontend/modules/mediapipe-pose/ios/MediaPipePoseView.swift
import ExpoModulesCore
import AVFoundation
import UIKit
import os

/// Logger para diagnostico visivel no Console.app em builds Release
private let viewLog = Logger(subsystem: "com.loadmanagerpro.mediapipe", category: "view")

/**
 * MediaPipePoseView — View nativa que renderiza camera + processa pose com MediaPipe.
 *
 * Camera preview via AVCaptureSession.
 * Deteccao de pose via PoseLandmarkerService (MediaPipe Tasks Vision, liveStream mode).
 * JS recebe apenas landmarks serializados (struct leve).
 *
 * REGRAS:
 * - MediaPipe NUNCA roda no main thread
 * - Camera pipeline NUNCA e bloqueado (detectAsync e non-blocking)
 * - JS recebe dados reais, nao simulacao
 */
class MediaPipePoseView: ExpoView, AVCaptureVideoDataOutputSampleBufferDelegate {

  // MARK: - Event Dispatchers
  private let onPoseDetected = EventDispatcher()
  private let onError = EventDispatcher()
  private let onCameraReady = EventDispatcher()

  // MARK: - Camera
  private let captureSession = AVCaptureSession()
  private let previewLayer = AVCaptureVideoPreviewLayer()
  private let videoOutput = AVCaptureVideoDataOutput()
  private let sessionQueue = DispatchQueue(label: "com.mediapipepose.session", qos: .userInitiated)
  private let processingQueue = DispatchQueue(label: "com.mediapipepose.processing", qos: .userInitiated)

  // MARK: - State
  private var isSessionRunning = false
  private var currentFacing: AVCaptureDevice.Position = .back
  private var isActive = false
  private var landmarkerService: PoseLandmarkerService?

  // MARK: - Config
  private var modelComplexity: Int = 0
  private var minDetectionConfidence: Float = 0.6
  private var minTrackingConfidence: Float = 0.6

  // MARK: - Init

  required init(appContext: AppContext? = nil) {
    super.init(appContext: appContext)
    viewLog.info("[View] CREATED")
    setupPreviewLayer()
  }

  // MARK: - Layout

  override func layoutSubviews() {
    super.layoutSubviews()
    previewLayer.frame = bounds
  }

  // MARK: - Setup

  private func setupPreviewLayer() {
    previewLayer.session = captureSession
    previewLayer.videoGravity = .resizeAspectFill
    layer.addSublayer(previewLayer)
  }

  private func initializeLandmarker() {
    viewLog.info("[View] initializeLandmarker START")
    landmarkerService = PoseLandmarkerService(
      modelComplexity: modelComplexity,
      minDetectionConfidence: minDetectionConfidence,
      minTrackingConfidence: minTrackingConfidence,
      onResult: { [weak self] landmarks, timestampMs, width, height in
        self?.dispatchPoseEvent(landmarks: landmarks, timestampMs: timestampMs, width: width, height: height)
      },
      onError: { [weak self] message in
        viewLog.error("[View] onError callback: \(message, privacy: .public)")
        self?.onError(["message": message])
      }
    )
    let avail = landmarkerService?.isAvailable == true
    viewLog.info("[View] initializeLandmarker END - isAvailable=\(avail, privacy: .public)")
  }

  // MARK: - Props setters

  func setCameraFacing(_ facing: String) {
    let newPosition: AVCaptureDevice.Position = facing == "front" ? .front : .back
    guard newPosition != currentFacing else { return }
    currentFacing = newPosition

    if isSessionRunning {
      sessionQueue.async { [weak self] in
        self?.reconfigureCamera()
      }
    }
  }

  func setIsActive(_ active: Bool) {
    viewLog.info("[View] setIsActive: \(active, privacy: .public) (was: \(self.isActive, privacy: .public))")
    guard active != isActive else { return }
    isActive = active

    if active {
      startSession()
    } else {
      stopSession()
    }
  }

  func setModelComplexity(_ complexity: Int) {
    modelComplexity = complexity
  }

  func setMinDetectionConfidence(_ confidence: Float) {
    minDetectionConfidence = confidence
  }

  func setMinTrackingConfidence(_ confidence: Float) {
    minTrackingConfidence = confidence
  }

  // MARK: - Session Lifecycle

  private func startSession() {
    viewLog.info("[View] startSession CALLED")
    sessionQueue.async { [weak self] in
      guard let self = self else { return }

      viewLog.info("[View] startSession EXECUTING on sessionQueue")
      self.initializeLandmarker()
      self.configureSession()

      if !self.captureSession.isRunning {
        viewLog.info("[View] captureSession.startRunning()")
        self.captureSession.startRunning()
        self.isSessionRunning = true

        DispatchQueue.main.async {
          viewLog.info("[View] onCameraReady dispatched")
          self.onCameraReady([:])
        }
      }
    }
  }

  private func stopSession() {
    viewLog.info("[View] stopSession CALLED")
    sessionQueue.async { [weak self] in
      guard let self = self else {
        viewLog.warning("[View] stopSession - self is nil (already deallocated)")
        return
      }

      viewLog.info("[View] stopSession EXECUTING - isRunning=\(self.captureSession.isRunning, privacy: .public)")
      if self.captureSession.isRunning {
        self.captureSession.stopRunning()
        self.isSessionRunning = false
        viewLog.info("[View] captureSession STOPPED")
      }

      self.landmarkerService?.close()
      self.landmarkerService = nil
      viewLog.info("[View] landmarkerService CLOSED")
    }
  }

  private func configureSession() {
    captureSession.beginConfiguration()
    captureSession.sessionPreset = .medium

    captureSession.inputs.forEach { captureSession.removeInput($0) }
    captureSession.outputs.forEach { captureSession.removeOutput($0) }

    guard let device = AVCaptureDevice.default(
      .builtInWideAngleCamera,
      for: .video,
      position: currentFacing
    ),
    let input = try? AVCaptureDeviceInput(device: device),
    captureSession.canAddInput(input) else {
      viewLog.error("[View] ERRO: camera nao disponivel")
      DispatchQueue.main.async { [weak self] in
        self?.onError(["message": "Camera nao disponivel para posicao selecionada"])
      }
      captureSession.commitConfiguration()
      return
    }

    captureSession.addInput(input)

    videoOutput.alwaysDiscardsLateVideoFrames = true
    videoOutput.videoSettings = [
      kCVPixelBufferPixelFormatTypeKey as String: kCVPixelFormatType_32BGRA
    ]
    videoOutput.setSampleBufferDelegate(self, queue: processingQueue)

    if captureSession.canAddOutput(videoOutput) {
      captureSession.addOutput(videoOutput)

      if let connection = videoOutput.connection(with: .video) {
        if #available(iOS 17.0, *) {
          if connection.isVideoRotationAngleSupported(90) {
            connection.videoRotationAngle = 90
          }
        } else {
          if connection.isVideoOrientationSupported {
            connection.videoOrientation = .portrait
          }
        }
        connection.isVideoMirrored = (currentFacing == .front)
      }
    }

    captureSession.commitConfiguration()
    viewLog.info("[View] configureSession DONE")
  }

  private func reconfigureCamera() {
    captureSession.beginConfiguration()

    captureSession.inputs.forEach { captureSession.removeInput($0) }

    guard let device = AVCaptureDevice.default(
      .builtInWideAngleCamera,
      for: .video,
      position: currentFacing
    ),
    let input = try? AVCaptureDeviceInput(device: device),
    captureSession.canAddInput(input) else {
      captureSession.commitConfiguration()
      return
    }

    captureSession.addInput(input)

    if let connection = videoOutput.connection(with: .video) {
      connection.isVideoMirrored = (currentFacing == .front)
    }

    captureSession.commitConfiguration()
  }

  // MARK: - AVCaptureVideoDataOutputSampleBufferDelegate

  func captureOutput(
    _ output: AVCaptureOutput,
    didOutput sampleBuffer: CMSampleBuffer,
    from connection: AVCaptureConnection
  ) {
    guard isActive, let landmarkerService = landmarkerService else { return }

    // Timestamp monotônico do frame (CMSampleBuffer PTS)
    let timestampCM = CMSampleBufferGetPresentationTimeStamp(sampleBuffer)
    let timestampInMs = Int(CMTimeGetSeconds(timestampCM) * 1000)

    landmarkerService.detect(sampleBuffer: sampleBuffer, timestampMs: timestampInMs)
  }

  // MARK: - Event Dispatch

  private func dispatchPoseEvent(landmarks: [[String: Any]], timestampMs: Int, width: Int, height: Int) {
    DispatchQueue.main.async { [weak self] in
      self?.onPoseDetected([
        "landmarks": landmarks,
        "timestamp": timestampMs,
        "frameWidth": width,
        "frameHeight": height,
      ])
    }
  }

  // MARK: - Cleanup

  deinit {
    viewLog.info("[View] DEINIT - cleanup sincrono")

    // 1. Remover delegate IMEDIATAMENTE para prevenir entrega de frames
    //    a um objeto ja desalocado (causa raiz do crash)
    videoOutput.setSampleBufferDelegate(nil, queue: nil)

    // 2. Capturar referencias locais (self nao pode ser usado apos deinit)
    let session = captureSession
    let service = landmarkerService
    let queue = sessionQueue

    // 3. Dispatch cleanup para sessionQueue com referencias fortes
    queue.async {
      viewLog.info("[View] DEINIT cleanup - stopping session")
      if session.isRunning {
        session.stopRunning()
      }
      service?.close()
      viewLog.info("[View] DEINIT cleanup DONE")
    }
  }
}

File: frontend/modules/mediapipe-pose/android/src/main/java/expo/modules/mediapipepose/MediaPipePoseView.kt
package expo.modules.mediapipepose

import android.content.Context
import android.widget.FrameLayout
import androidx.camera.core.CameraSelector
import androidx.camera.core.ImageAnalysis
import androidx.camera.core.Preview
import androidx.camera.lifecycle.ProcessCameraProvider
import androidx.camera.view.PreviewView
import androidx.core.content.ContextCompat
import androidx.lifecycle.LifecycleOwner
import expo.modules.kotlin.AppContext
import expo.modules.kotlin.viewevent.EventDispatcher
import expo.modules.kotlin.views.ExpoView
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

/**
 * MediaPipePoseView — View nativa Android com CameraX + MediaPipe Pose Landmarker.
 *
 * Arquitetura:
 * - CameraX gerencia a camera (preview via PreviewView + analise via ImageAnalysis)
 * - PoseLandmarkerService processa frames em thread separado (VIDEO mode, sincrono)
 * - Resultados sao enviados ao JS via EventDispatcher (onPoseDetected)
 *
 * REGRAS:
 * - MediaPipe NUNCA roda no main thread
 * - Camera pipeline NUNCA e bloqueado (STRATEGY_KEEP_ONLY_LATEST)
 * - JS recebe apenas landmarks serializados (Map leve)
 */
class MediaPipePoseView(context: Context, appContext: AppContext) : ExpoView(context, appContext) {

  // Event dispatchers para o JS
  private val onPoseDetected by EventDispatcher()
  private val onError by EventDispatcher()
  private val onCameraReady by EventDispatcher()

  // Camera
  private val previewView: PreviewView = PreviewView(context).apply {
    layoutParams = FrameLayout.LayoutParams(
      FrameLayout.LayoutParams.MATCH_PARENT,
      FrameLayout.LayoutParams.MATCH_PARENT
    )
    implementationMode = PreviewView.ImplementationMode.COMPATIBLE
  }

  private val analysisExecutor: ExecutorService = Executors.newSingleThreadExecutor()
  private var cameraProvider: ProcessCameraProvider? = null
  private var landmarkerService: PoseLandmarkerService? = null

  // State
  private var currentFacing = CameraSelector.LENS_FACING_BACK
  private var isActive = false
  private var isCameraBound = false

  // Config
  private var modelComplexity: Int = 0
  private var minDetectionConfidence: Float = 0.6f
  private var minTrackingConfidence: Float = 0.6f

  init {
    addView(previewView)
  }

  // MARK: - Props setters

  fun setCameraFacing(facing: String) {
    val newFacing = if (facing == "front") CameraSelector.LENS_FACING_FRONT else CameraSelector.LENS_FACING_BACK
    if (newFacing != currentFacing) {
      currentFacing = newFacing
      if (isCameraBound) {
        rebindCamera()
      }
    }
  }

  fun setIsActive(active: Boolean) {
    if (active != isActive) {
      isActive = active
      if (active) {
        startCamera()
      } else {
        stopCamera()
      }
    }
  }

  fun setModelComplexity(complexity: Int) {
    modelComplexity = complexity
  }

  fun setMinDetectionConfidence(confidence: Float) {
    minDetectionConfidence = confidence
  }

  fun setMinTrackingConfidence(confidence: Float) {
    minTrackingConfidence = confidence
  }

  // MARK: - Camera Lifecycle

  private fun startCamera() {
    val cameraProviderFuture = ProcessCameraProvider.getInstance(context)

    cameraProviderFuture.addListener({
      try {
        cameraProvider = cameraProviderFuture.get()
        initializeLandmarker()
        bindCamera()
      } catch (e: Exception) {
        onError(mapOf("message" to "Falha ao iniciar camera: ${e.message}"))
      }
    }, ContextCompat.getMainExecutor(context))
  }

  private fun stopCamera() {
    cameraProvider?.unbindAll()
    isCameraBound = false
    landmarkerService?.close()
    landmarkerService = null
  }

  private fun rebindCamera() {
    cameraProvider?.unbindAll()
    isCameraBound = false
    bindCamera()
  }

  private fun initializeLandmarker() {
    landmarkerService?.close()
    landmarkerService = PoseLandmarkerService(
      context = context,
      modelComplexity = modelComplexity,
      minDetectionConfidence = minDetectionConfidence,
      minTrackingConfidence = minTrackingConfidence,
      onResult = { landmarks, timestampMs, width, height ->
        // Dispatch para o JS no main thread
        post {
          onPoseDetected(mapOf(
            "landmarks" to landmarks,
            "timestamp" to timestampMs,
            "frameWidth" to width,
            "frameHeight" to height
          ))
        }
      },
      onError = { message ->
        post {
          onError(mapOf("message" to message))
        }
      }
    )
  }

  private fun bindCamera() {
    val provider = cameraProvider ?: return
    val lifecycleOwner = getLifecycleOwner() ?: run {
      onError(mapOf("message" to "LifecycleOwner nao disponivel"))
      return
    }

    try {
      provider.unbindAll()

      val cameraSelector = CameraSelector.Builder()
        .requireLensFacing(currentFacing)
        .build()

      // Preview
      val preview = Preview.Builder()
        .build()
        .also {
          it.surfaceProvider = previewView.surfaceProvider
        }

      // Image Analysis para frames
      val imageAnalysis = ImageAnalysis.Builder()
        .setBackpressureStrategy(ImageAnalysis.STRATEGY_KEEP_ONLY_LATEST)
        .setOutputImageFormat(ImageAnalysis.OUTPUT_IMAGE_FORMAT_RGBA_8888)
        .build()

      imageAnalysis.setAnalyzer(analysisExecutor) { imageProxy ->
        if (isActive && landmarkerService != null) {
          landmarkerService?.detect(imageProxy)
        } else {
          imageProxy.close()
        }
      }

      provider.bindToLifecycle(
        lifecycleOwner,
        cameraSelector,
        preview,
        imageAnalysis
      )

      isCameraBound = true
      onCameraReady(mapOf<String, Any>())

    } catch (e: Exception) {
      onError(mapOf("message" to "Falha ao vincular camera: ${e.message}"))
    }
  }

  private fun getLifecycleOwner(): LifecycleOwner? {
    // Tenta obter do contexto (Activity)
    var ctx = context
    while (ctx != null) {
      if (ctx is LifecycleOwner) return ctx
      ctx = if (ctx is android.content.ContextWrapper) ctx.baseContext else null
    }
    return null
  }

  // MARK: - Cleanup

  override fun onDetachedFromWindow() {
    super.onDetachedFromWindow()
    stopCamera()
    analysisExecutor.shutdown()
  }
}
